platform/pc/service.c

```diff
@@ -395,3 +422,69 @@ void init_service(u64 rdi, u64 rsi)
     *(u64 *)stack_location = 0;
     switch_stack(stack_location, init_service_new_stack);
 }
+
+extern boolean init_hpet(kernel_heaps kh);
+
+void detect_hypervisor(kernel_heaps kh)
+{
+    if (!kvm_detect(kh)) {
+        init_debug("probing for Xen hypervisor");
+        if (!xen_detect(kh)) {
+            if (!hyperv_detect(kh)) {
+                init_debug("no hypervisor detected; assuming qemu full emulation");
+                if (!init_hpet(kh)) {
+                    halt("HPET initialization failed; no timer source\n");
+                }
+            } else {
+                init_debug("hyper-v hypervisor detected");
+            }
+        } else {
+            init_debug("xen hypervisor detected");
+        }
+    } else {
+        init_debug("KVM detected");
+    }
+}
+
+void detect_devices(kernel_heaps kh, storage_attach sa)
+{
+    /* Probe for PV devices */
+    if (xen_detected()) {
+        init_debug("probing for Xen PV network...");
+        init_xennet(kh);
+        init_xenblk(kh, sa);
+        status s = xen_probe_devices();
+        if (!is_ok(s))
+            halt("xen probe failed: %v\n", s);
+        return;
+    }
+
+    if (hyperv_detected()) {
+        boolean hyperv_storvsc_attached = false;
+        init_debug("probing for Hyper-V PV network...");
+        init_vmbus(kh);
+        status s = hyperv_probe_devices(sa, &hyperv_storvsc_attached);
+        if (!is_ok(s))
+            halt("Hyper-V probe failed: %v\n", s);
+        if (!hyperv_storvsc_attached)
+            init_ata_pci(kh, sa); /* hvm ata fallback */
+        return;
+    }
+
+    init_debug("hypervisor undetected or HVM platform; registering all PCI drivers...");
+
+    /* net */
+    init_virtio_network(kh);
+    init_vmxnet3_network(kh);
+    init_aws_ena(kh);
+
+    /* storage */
+    init_virtio_blk(kh, sa);
+    init_virtio_scsi(kh, sa);
+    init_pvscsi(kh, sa);
+    init_nvme(kh, sa);
+    init_ata_pci(kh, sa);
+
+    /* misc / platform */
+    init_acpi(kh);
```

Comment:
init_acpi() should be called regardless of hypervisor detection. Not calling it means not registering the PIIX4 powerdown handler, which is needed in different x86_64 cloud instance types (including AWS Xen instances) to stop an instance when the program terminates.

Reply:
Ok, I made the call to init_acpi() unconditional.